### ISSUE TITLE
[Q] Clarification on the implementation of hyperband algorithm in Sweeps

### ISSUE THREAD

**Issue by scho97** [wandb/client#2771]
Hello,

I was reading the [documentation](https://docs.wandb.ai/guides/sweeps/configuration#hyperband) about the Hyperband algorithm in Sweeps. In the original paper, each bracket refers to an implementation of Successive Halving Algorithm (SHA). However, in the documentation, it says that a bracket is one pre-set iteration counts. I am confused what the iteration is referring to and how the Hyperband works in sweeps.

In the cases where I have `max_iter=150`, `eta=3`, and `s=4`, the early stopping will occur based on the epochs [150/3, 150/9, 150/27, 150/81] and the thresholds calculated based on these epochs, as seen in the [source code](https://github.com/wandb/client/blob/master/wandb/sweeps/hyperband_stopping.py#L186-L195).

My question is that when I looked at the code, I find SHA is applied only once rather than multiple times. How is SHA implemented in the W&B version of Hyperband?

Also, is the hyperband implemented serially or parallelly? 
If serial and the history of different runs are necessary to decide early stopping, does it mean that the first run will never be pruned?
And if it is parallel such that every run is ran for a set epoch and then compared, how does this fit with Bayesian Gaussian processes which require the hyperparameters in each run to be sampled from previous run results?

Thank you.

**Comment by anmolmann** [comment]
@scho97 apologies for the delay here.
Our implementation is based on the "distributed implementation" idea briefly discussed in the conclusion section of the paper. And, `eta`, and `s` define the brackets when hyperband algo should consider stopping runs based on all the other runs at that same iteration number. 

Also, in our default implementation of starting & running sweeps, the early stopping criteria is serialized and not parallelized. However, you can do some configurations on your end and launch multiple runs in parallel yourself. Here's [our guide](https://stackoverflow.com/questions/63469762/weightsbiases-sweep-keras-k-fold-validation) on how to launch multiple runs from a sweep.

So, once the first run logs `your_optimization_metric` [150/3] times, the first bracket is setup. Now, the next run when it logs `your_optimization_metric` 50 times, it will be compared to this bracket, and similarly at [150/9] iterations & for all future runs. 
A good way to see what is happening is to run in "local" mode with full debugging: `wandb sweep --controller --verbose sweep-grid-hyperband.yaml`. This will give you insight into how your bands are setup and what they are set at.
Finally, this means that our sweep monitor service will monitor all your runs, and it will stop any run where the metric you are optimizing (specified in your `sweep-grid-hyperband.yaml` file) is in the bottom half of all other runs which reported that metric at the same iteration. When you run the sweep controller locally, it will show you what those bands are computed to as the sweep is executed.

**Comment by anmolmann** [comment]
Hi @scho97, I wanted to follow up on this request. Please let us know if we can be of further assistance or if your issue has been resolved.

**Comment by scho97** [comment]
Hi @anmolmann, thank you for your reply.
Firstly, I thought the first bracket sets up once the first run logs the metric [150/81] times, as the bands (i.e., `bands=[150/3, 150/9, 150/27, 150/81]` here) are sorted.

I am still confused how the idea of bracket in the paper fits into the ones implemented in `wandb`. Are you saying there will be `n=len(bands)` brackets or there will be one bracket in which each run is compared at [150/81] iterations, [150/27] iterations, and so on.

As for the "local" mode, I will try it out and monitor how the bands are setup. Thank you.

**Comment by anmolmann** [comment]
@scho97 Brackets are configured at static iterations for a specified metric (where an iteration is the number of times a metric has been logged — if the metric is logged every epoch, then there are epoch iterations). So, there will be 4 brackets in this case: `[150/3, 150/9, 150/27, 150/81]` and each bracket will have respective bands (`max_iter/eta`). So, at a specific iteration: active runs will be compared against all other runs at that iteration to determine which should be stopped.

Also, did you try out the `local` mode and see how the bands are setup in your case?

**Comment by anmolmann** [comment]
Hi @scho97, I wanted to follow up on this request. Please let us know if we can be of further assistance or if your issue has been resolved.

**Comment by tjades** [comment]
Hi, following up on this, the wand implementation is just very unrelated to the algorithm proposed in the hyperband paper, right?
The paper addresses the tradeoff between "how many parameter settings can i try" and "how long can I evaluate each setting".
The wandb implementation, however, doesn't address the "how many parameter settings can i try" question, but only provides checkpoints at which early stopping is applied with regards to earlier runs. Therefore, the concept of brackets in wandb-hyperband and actual hyperband is something else, correct?
